java/flight/src/main/java/org/apache/arrow/flight/FlightProducer.java

```diff
@@ -181,5 +183,11 @@ void doAction(CallContext context, Action action,
 
     /** Whether the call has been cancelled by the client. */
     boolean isCancelled();
+
+    /** Get the middleware instance of the given type for this call. */
+    <T extends FlightServerMiddleware> T getMiddleware(FlightServerMiddleware.Key<T> key);
+
+    /** Get an immutable map of middleware for this call. */
```

Comment:
is there a way to make the contract of immutability clear based on the return type?

Reply:
Not without putting a Guava type in the public API, as Java collections don't have a separate type for immutable maps.